### DISCUSSION TITLE
常见问题（FAQ）

### DISCUSSION THREAD

**Discussion by zengbin93** [waditu/czsc#32] [Q&A]
非常感谢大家的支持！关于缠中说禅 `czsc` 的任何问题都可以在这里提出来，提问之前最好还是把已有的问题看一遍，也许你的疑问能够从以前的问题中得到答案。

**Comment by 450916608** [comment]
可不可以占个位置，表示下对群主的敬意和支持，感谢开源分享并提供技术指导！

**Comment by searchlink** [comment]
关于factors.py中的一些疑问：
```
c1.bi_list[-1].direction == Direction.Down and len(c1.bars_ubi) <= 7
c1.bi_list[-1].direction == Direction.Up and len(c1.bars_ubi) > 7
```
1. 未成笔包含的k线数跟7有什么关系怎么理解？为何要加这个限制？
2. 本级别最后一笔方向向下，次级别最后五笔构成三买，那么次级别最后一笔回踩中枢且不破中枢上沿 ,且创新高，有点不理解这里本级别和次级别的方向关系，为什么需要指定次级别形成三买的前提是本级别最后一笔方向向下？
```
    if c1.bi_list[-1].direction == Direction.Down and len(c1.bars_ubi) <= 7:    # 最后一笔方向向下
        factor_l1 = factor_key_base + "L1"  # L1 - 向下笔转折右侧
        # L1A
        if c2.signals['倒1五笔'] in [Signals.X5LB0.value, Signals.X5LB1.value]:    # 最后五笔形成三买
            if c2_h9 == c2.bi_list[-1].high:    # 近九笔高点等于近五笔最后一笔高点
                v = Factors['{}A1'.format(factor_l1)].value     # 本级别
            elif "顶背弛" in c1.signals['倒2五笔'] \
                    and "顶背弛" in c1.signals['倒2七笔'] \
                    and "顶背弛" in c1.signals['倒2九笔']:
                v = Factors['{}A2'.format(factor_l1)].value
            elif "底背弛" in c1.signals['倒1五笔'] \
                    and "底背弛" in c1.signals['倒1七笔'] \
                    and "底背弛" in c1.signals['倒1九笔']:
                v = Factors['{}A3'.format(factor_l1)].value
            else:
                v = Factors['{}A0'.format(factor_l1)].value
```
3. 得到关于多种不同级别的因子信号后，关于怎么使用起来这些因子信号来参与介入？

**Reply by zengbin93** [reply]
回答如下，仅供参考：

1. 在缠论中，笔的基础形态至少有7根无包含K线，7这个数字在这里有特殊含义，通常的理解是一个笔延伸已经有了7根无包含K线，那就有了笔的基础形态。这里主要用来区分笔延伸和转折的方向。

2. 这是为了提高三买的胜率，就是只做本级别向上笔的第一个三买，所以限定本级别最后一个能够在图上画出来的笔是向下的。

3. 每个因子信号都需要在个股的历史数据上验证，每个股票适应的因子可能都是不一样的。另外一种验证方法是在全市场每天跟踪一些你自己认为很有价值的因子。这里见仁见智，高下在此区分开！一个更好的方法是把你平常看盘的经验写成因子，然后验证，慢慢的就能形成自己的体系。

**Reply by searchlink** [reply]
继续打扰大佬了~还是继续扔下两个问题：
1. 就是关于两笔的力度比较。代码中是引入用一笔的(最高价-最低价)来进行力度比较，想问的是考虑用价差比较而不用macd比较，处于什么考虑，哪个更准，还是如禅师提到技术指标时说到只考虑均线和成交量比其他技术指标都有意义；
2. 关于拟合优度的引入，度量一笔和很多散点的拟合程度，在判断信号中引入的目的是啥？

**Reply by zengbin93** [reply]
回答如下，仅供参考：

1. 笔的力度衡量可以用很多方法，MACD是缠师用的，我这里用价格差来衡量主要是为了提高计算效率；实际应用发现，用价差衡量力度发现的背驰跟MACD的基本上没有差别，对此有疑问的，可以自行验证。

2. 拟合优度之类的非形态类信号，主要是为了做机器学习模型准备的，也有些会用来辅助形态判断，提高胜率。

**Reply by Xi40Yu** [reply]
1.关于一笔有7个K先的问题我有点不同的看法。就图形上来说，一笔是7个K线，但实际中，因为走势是连续的，所以一笔最少5个K线即可。也就是分型中间的K线，分型右边的K线，然后是独立K线（新笔的规则可以没有这根），第二个分型左边的K线，第二个分型中间的K线，一共5个。同理，一条线段就是13个K线。

2.力度的比较，缠给过一个方法，就是当一个上涨笔延伸的过程中，如果出现顶分型，不发生转折，而后继续上涨延伸，则这个延伸过程中的顶分型可以作为分界线，比较这一笔内部的相对力度。这种方法通常用在大周期上，希望可以后续加上。有了这个方法，一来可以在大周期上选择趋势向好的股票，比如刚刚构筑完底分型，新的上涨一笔延伸中的股票。便于在大级别上比较走势力度。

**Reply by alxbj** [reply]
> 回答如下，仅供参考：
> 
> 1. 在缠论中，笔的基础形态至少有7根无包含K线，7这个数字在这里有特殊含义，通常的理解是一个笔延伸已经有了7根无包含K线，那就有了笔的基础形态。这里主要用来区分笔延伸和转折的方向。
> 2. 这是为了提高三买的胜率，就是只做本级别向上笔的第一个三买，所以限定本级别最后一个能够在图上画出来的笔是向下的。
> 3. 每个因子信号都需要在个股的历史数据上验证，每个股票适应的因子可能都是不一样的。另外一种验证方法是在全市场每天跟踪一些你自己认为很有价值的因子。这里见仁见智，高下在此区分开！一个更好的方法是把你平常看盘的经验写成因子，然后验证，慢慢的就能形成自己的体系。

我记得7根k线是顶底分型的6根+中间的1根=7根=1笔的基础形态

**Comment by choastrade** [comment]
看到新版本，只有笔，没有段，这是出于怎么考虑呢？

**Reply by zengbin93** [reply]
笔和线段都可以作为 a0 ，在禅师的原文中，前60课基本上都是以笔作为 a0 的，后面才以线段作为 a0；相比于线段，笔作为 a0 有这么几个优势：1）在走势图上看起来更符合人眼直观观察，高低起伏明显；2）程序可以有更小的内存占用，因为一个线段走势包含的K线数量比一个笔包含的K线数量多好几倍。

虽然 `czsc` 这个库现在是以笔作为 a0，但是线段的结束的两种情况已经在形态信号中充分体现，线段相关内容并没有被丢弃。不严格的说，也可以认为以5分钟笔作为 a0 的交易等价于以1分钟线段作为 a0 的交易。

**Comment by Gavin-yzg** [comment]
有没有详细一点的教程,比如我想选日线三买的股票,实现思路 过程

**Reply by zengbin93** [reply]
三买在 `czsc` 这个库里面是一种特殊的五笔形态，你只需要传入K线到  `czsc.CZSC` 类中，所有的形态信号就会自动计算。

五笔形态识别的源码在这里，有兴趣可以看看：[https://github.com/zengbin93/czsc/blob/master/czsc/signals.py#L8](https://github.com/zengbin93/czsc/blob/master/czsc/signals.py#L8)

**Reply by Gavin-yzg** [reply]
好的,谢谢

**Comment by sunmichace** [comment]
您好。感谢开源
在factors.py中，发现一个小型可能是bug。
第39行中
`c2_l9 = max([x.high for x in c2.bi_list[-9:]])`
应该改为
`c2_l9 = max([x.low for x in c2.bi_list[-9:]])`

应该是写太快了：）

**Reply by zengbin93** [reply]
更新一下最新代码，这个问题之前就有人发现了，已经修正。

**Reply by searchlink** [reply]
继续打扰~问下函数_calculate_factors_d()，通过日线笔，来确定c1_, c2_, c3_
```
if len(c6.bars_ubi) > 9:    
    direction = Direction.Down if c6.bi_list[-1].direction == Direction.Up else Direction.Up    
    bis_c6_c4 = get_sub_span(c4.bi_list[-15:], start_dt=c6.bars_ubi[1].dt, end_dt=c6.bars_ubi[-1].dt, direction=direction)
```
1. 为何要判断未成笔且无包含关系的k线数大于9？ 
```
if 9 >= len(bis_c6_c4) >= 5:        
    c1_, c2_, c3_ = c6, c4, c2
elif len(bis_c6_c4) > 9:            
    c1_, c2_, c3_ = c6, c5, c3
elif len(bis_c6_c4) < 5:
    c1_, c2_, c3_ = c6, c3, c1
else:
    raise ValueError
```
2.  获取了日线下未成笔且无包含关系的30分钟笔之后，这里通过判断笔数，来确定c1_, c2_, c3_，不明白9 >= len(bis_c6_c4) >= 5这个为什么对应日线、30分钟和5分钟，len(bis_c6_c4) > 9对应日线、60分钟和15分钟，以及
len(bis_c6_c4) < 5对应日线、15分钟和1分钟。
3. 关于函数_calculate_factors_f60(), 同样的问题。获取了60分钟未成笔且无包含关系的15分钟笔之后，这里通过判断笔数，不明白通过笔数大于9，len(c5.bars_ubi) > 9来确定c1_, c2_, c3_，不明白len(bis_c5_c4) >= 5这个为什么对应60分钟、15分钟和1分钟，以及 len(bis_c5_c4) < 5对应60分钟、5分钟和1分钟。
```
if len(c5.bars_ubi) > 9:
    direction = Direction.Down if c5.bi_list[-1].direction == Direction.Up else Direction.Up
   bis_c5_c4 = get_sub_span(c3.bi_list[-15:], start_dt=c5.bars_ubi[1].dt, end_dt=c5.bars_ubi[-1].dt, direction=direction)

if len(bis_c5_c4) >= 5:
   c1_, c2_, c3_ = c5, c3, c1
 else:
   c1_, c2_, c3_ = c5, c2, c1
```

**Reply by sunmichace** [reply]
这个问题，我看的时候也思考过。
按照大佬的模型，主要将笔按数量分类形态，分别是5,7,9。所以9和5是这么来的。
其次，对应关系，如果子区间的30m笔数量大于9对应的是日，60,15。这是组合中除了日，60,30后最大的关系（日，60,15）。算是顶格的本级别，次级别，子级别了。其他的类似这么理解。
但究竟为什么这么设计?
我估计还是个人喜好吧，毕竟按自己模型来。

**Reply by zengbin93** [reply]
基本上就是 @sunmichace 理解的意思。这里使用的一个前提假设是：本级别笔对应次级别走势应该有5~9笔，小于5笔就降一级，大于9笔就升一级。这是我个人的理解，算是一个走势分解的例子吧，有兴趣的可以按自己的理解来实现新的因子。

**Comment by zoaamour** [comment]
我其实想先回测一下factor和event的效果如何，但是不知道如何下手。
问题是：1、有产生历史信号的代码吗，提示一下在哪里就好了。
已经在看代码，有不少想请教的地方。为表示尊敬，找个发言多不显眼的地方提问。
但是我也不肯定自己问的能否算常见问题，没有时间回答就不用理我好了。

**Reply by zengbin93** [reply]
@zoaamour `scripts_gm` 下面有策略案例

**Comment by xiduofu** [comment]
您好，今天刚搜到您的开源项目，感谢分享。
但是目前无从下手，不太熟悉整体架构，该如何调用。
也看到您csdn上关于 tushare的demo了，最新版本下已无法执行，请问您有最新版的demo或是文档可以分享吗，谢谢！

**Comment by 450916608** [comment]
这是一封自动回复邮件。已经收到您的来信，我会尽快回复。

**Comment by aligame5288** [comment]
请问一下“ czsc-0.10.11” 为什么只能获取到分型、笔 的数据，获取不到 线段和中枢的数据呢

**Comment by 450916608** [comment]
这是一封自动回复邮件。已经收到您的来信，我会尽快回复。
